channeldb/invoices.go

```diff
@@ -1831,18 +1818,49 @@ func (d *DB) updateInvoice(hash lntypes.Hash, invoices,
 					key)
 			}
 
-			err := cancelSingleHtlc(now, htlc, invoice.State)
+			err := cancelSingleHtlc(now, htlc, newState)
 			if err != nil {
 				return nil, err
 			}
 
 			// Delete processed cancel action, so that we can check
 			// later that there are no actions left.
 			delete(cancelHtlcs, key)
+		}
+	}
 
-			continue
+	// Verify that we didn't get an action for htlcs that are not present on
+	// the invoice.
+	if len(cancelHtlcs) > 0 {
+		return nil, errors.New("cancel action on non-existent htlc(s)")
+	}
+
+	// At this point, the set of accepted HTLCs should be fully
+	// populated with added HTLCs or removed of canceled ones. Update
+	// invoice state if the update descriptor indicates an invoice state
+	// change, which depends on having an accurate view of the accepted
+	// HTLCs.
+	if update.State != nil {
+		err := updateInvoiceState(&invoice, hash, *update.State)
+		if err != nil {
+			return nil, err
+		}
+
+		if update.State.NewState == ContractSettled {
+			err := setSettleMetaFields(
+				settleIndex, invoiceNum, &invoice, now,
+			)
+			if err != nil {
+				return nil, err
+			}
 		}
+	}
 
+	// With any invoice level state transitions recorded, we'll now finalize
+	// the process by updating the state transitions for individual HTLCs
+	// and recalculate the total amount paid to the invoice.
+	var amtPaid lnwire.MilliSatoshi
+	for _, htlc := range invoice.Htlcs {
```

Comment:
don't we still have to check that the HTLC is not among the canceled ones?

Reply:
The invoice being updated? That's still being checked above when we go to cancel a single HTLC.

Reply:
Earlier we would iterate through all the invoice's HTLCs, calling `cancelSingleHtlc` on those canceled by the update, and calling `updateHtlc` on the rest.

With this commit we are still calling `cancelSingleHtlc` on canceled HTLCs, but we are calling `updateHtlc` for ALL, regardless of canceled or non-canceled.

I'm wondering if this is an intentional  change.

Reply:
it probably isn't necessary to call it on all HTLCs, but it will be a NOP for anything that was previously canceled since all of the htlc state changes are gated by checks on `htlc.State == HtlcStateAccepted`